src/java/org/apache/cassandra/db/compaction/BackgroundCompactionRunner.java

```diff
@@ -114,166 +114,150 @@
+     *         completed by the caller.
+     */
+    CompletableFuture<RequestResult> requestCompaction(ColumnFamilyStore cfs)
+    {
+        logger.trace("Requested background compaction for {}.{}", cfs.getKeyspaceName(), cfs.getTableName());
+        FutureRequestResult p = compactionRequests.computeIfAbsent(cfs, ignored -> new FutureRequestResult());
+        if (!maybeScheduleNextCheck())
+        {
+            logger.info("Executor has been shut down, background compactions check will not be scheduled");
+            p.completeInternal(RequestResult.ABORTED);
         }
+        return p;
     }
 
     void shutdown()
     {
         checkExecutor.shutdown();
-        compactionRequests.values().forEach(promise -> promise.cancel(false));
+        compactionRequests.values().forEach(promise -> promise.completeInternal(RequestResult.ABORTED));
+        // it's okay to complete a CompletableFuture more than one on race between request, run and shutdown
     }
 
-    int getOngoingCompactionsCount(ColumnFamilyStore cfs)
+    int getOngoingCompactionsCount()
     {
-        return ongoingCompactions.count(cfs);
+        return ongoingCompactions.get();
     }
 
     @Override
     public void run()
     {
         logger.trace("Running background compactions check");
 
+        // When the executor is fully occupied, we delay acting on this request until a thread is available. This
+        // helps make a better decision what exactly to compact (e.g. if we issue the request now we may select n
+        // sstables, while by the time this request actually has a thread to execute on more may have accumulated
+        // and it may be better to compact all).
+        // Note that we make a request whenever a task completes and thus this method is guaranteed to run again
+        // when threads free up.
+        if (ongoingCompactions.get() >= compactionExecutor.getMaximumPoolSize())
+        {
+            logger.trace("Background compaction threads are busy; delaying new compactions check until there are free threads");
+            return;
+        }
+
         // We shuffle the CFSs for which the compaction was requested so that with each run we traverse those CFSs
         // in different order and make each CFS have equal chance to be selected
         ArrayList<ColumnFamilyStore> compactionRequestsList = new ArrayList<>(compactionRequests.keySet());
         Collections.shuffle(compactionRequestsList, random);
 
         for (ColumnFamilyStore cfs : compactionRequestsList)
         {
-            CompletableFuture<RequestResult> promise = compactionRequests.remove(cfs);
+            if (ongoingCompactions.get() >= compactionExecutor.getMaximumPoolSize())
+            {
+                logger.trace("Background compaction threads are busy; delaying new compactions check until there are free threads");
+                return;
```

Comment:
shouldn't we reschedule a next check, perhaps with slight delay before returning?